File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {

	static int[] dx = {-2, -1, 1, 2, 2, 1, -1, -2};
	static int[] dy = {1, 2, 2, 1, -1, -2, -2 ,-1};

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine());

		for (int i = 0; i < t; i++) {
			int l = Integer.parseInt(br.readLine());;

			StringTokenizer st = new StringTokenizer(br.readLine());
			int startX = Integer.parseInt(st.nextToken());
			int startY = Integer.parseInt(st.nextToken());

			st = new StringTokenizer(br.readLine());
			int endX = Integer.parseInt(st.nextToken());
			int endY = Integer.parseInt(st.nextToken());

			System.out.println(bfs(l, startX, startY, endX, endY));
		}
	}

	private static int bfs(int l, int startX, int startY, int endX, int endY) {
		ArrayDeque<int[]> queue = new ArrayDeque();
		boolean[][] visited = new boolean[l][l];
		queue.offer(new int[]{startX, startY, 0});
		visited[startX][startY] = true;

		while (!queue.isEmpty()) {
			int[] current = queue.poll();
			int x = current[0];
			int y = current[1];
			int count = current[2];

			if (x == endX && y == endY) {
				return count;
			}

			for (int i = 0; i < 8; i++) {
				int nx = x + dx[i];
				int ny = y + dy[i];

				if (0 <= nx && nx < l && 0 <= ny && ny < l && !visited[nx][ny]) {
					visited[nx][ny] = true;
					queue.offer(new int[]{nx, ny, count + 1});
				}
			}
		}
		return -1;
	}
}

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
import sys
from collections import deque

t = int(sys.stdin.readline().rstrip())
dx = [-2, -1, 1, 2, 2, 1, -1, -2]
dy = [1, 2, 2, 1, -1, -2, -2, -1]


def bfs(l, start_x, start_y, end_x, end_y):
    visited = [[False] * l for _ in range(l)]
    queue = deque()
    queue.append([start_x, start_y, 0])
    visited[start_x][start_y] = True

    while queue:
        current = queue.popleft()
        x = current[0]
        y = current[1]
        cnt = current[2]

        if x == end_x and y == end_y:
            return cnt

        for i in range(8):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx and nx < l and 0 <= ny and ny < l and not visited[nx][ny]:
                visited[nx][ny] = True
                queue.append([nx, ny, cnt + 1])

    return -1


for i in range(t):
    l = int(sys.stdin.readline().rstrip())

    start_x, start_y = map(int, sys.stdin.readline().rstrip().split())
    end_x, end_y = map(int, sys.stdin.readline().rstrip().split())

    print(bfs(l, start_x, start_y, end_x, end_y))
